aip/auth/4110.md

```diff
@@ -0,0 +1,165 @@
+
+- **GOOGLE_API_KEY**: [API keys][2] are a simple encrypted string that
+identifies an application, which are useful for accessing public data
+anonymously. The specified value of this environment variable **should** be used
+as the API key for the API requests. This value **must** be overridden by any
+user specified credentials at the programming level (e.g. through client
+options). And it **must** be ignored if GOOGLE_APPLICATION_CREDENTIALS is
+specified.
+
+- **GOOGLE_API_USE_CLIENT_CERTIFICATE:** The specified value **must** be
+either true or false. The client certificate **must** be ignored if this
+variable is set to false. The default value is false if the value is unset.
+
+```
+GOOGLE_API_USE_CLIENT_CERTIFICATE=[true|false]
+```
+
+### Inputs & Outputs
+
+From the input/output perspective, the inputs of _ADC_ **should** be the
+credentials as well as the underlying environment such as environment variables
+or metadata service that provides these credentials. 
+
+For example, the `GOOGLE_APPLICATION_CREDENTIALS` environment variable can provide
+the default credential JSON as the input here, or the well-known path that
+gCloud uses to store the default user credential JSON. The output is the access
+token that application can use to access the Google APIs. This access token
+__may__ be a bearer token, a certificate bound token or STS depending on the
+chosen authentication flow.
+
+## Expected Behavior
+
+This section outlines the expected behavior of the ADC. Auth libraries **must**
+implement these concepts in order to be considered complete.
+
+```graphviz
+digraph d_front_back {
+  rankdir=TB;
+  ranksep=0.3;
+  node [ style="filled,solid" shape=box fontname="Roboto" ];
+
+  check_env_var [ label="1. Check Environment Variables" ];
+  load_credentials [ label="2. Load Credentials" ];
+  check_metadata [ label="3. Check Metadata Service Credentials" ];
+  auth_flows [ label="4. Determine Auth Flows" ];
+  execute [ label="5. Execute Auth Flows" ];
+  sts [ label="6. Apply STS" ];
+
+  check_env_var -> load_credentials -> check_metadata ->auth_flows -> execute ->
+  sts;
+  load_credentials -> auth_flows;
+}
+```
+
+1. **Check environment variables**
+    1. Check GOOGLE_APPLICATION_CREDENTIALS
+        1. If set, go to step (2.2)
+    2. Check GOOGLE_API_KEY
+        1. If set, use the provided API key _[END]_
+    3. If neither of above is set, go to step (2)
+2. **Load credentials**
+    1. Check gcloud default credentials through its default path
+        1. If found go to step (2.2)
+        2. Otherwise go to step (3)
+    2. Check the provided credential type
+        1. If the credential is gcloud credentials, go to step (4)
+        2. If the credential is a service account key JSON, go to step (4)
+        3. If the credential is unknown type, return an error saying that _[END]_
+    3. Credentials not found _[END]_
+3. **Check Metadata Service Credentials**
```

Comment:
I think the steps under Load credentials and Metadata Service Credentials should be rearranged and simplified (and pointers in above sections updated). Several issues I spot: If gcloud default credentials is not found, and Metadata Service credentials is not found, we should not END, but rather look for credentials specified through client options, which is not mentioned here. Also after loading ANY credential files, we will go to step 4, so no need to break it into multiple checks.

If you'd like, I can take a stab at re-arranging the steps.

Reply:
client options specified credentials should have the highest priority, we will not enter ADC if client option is specified.